pyomo/repn/plugins/gams_writer.py

```diff
@@ -143,56 +231,79 @@ def _print_model( self, model, output_file, solver_capability, var_list,
         for con in model.component_data_objects(Constraint, active=True):
             if con.body.is_fixed():
                 continue
+
+            # Ensure GAMS will not encounter domain violations in presolver
+            # operations at current values, which are None (0) by default
+            # Used to handle log and log10 violations, for example
+            try:
+                value(con.body)
+            except:
+                raise RuntimeError("While evaluating: %s\n\tProblemWriter_gams"
+                                   " encountered an error. Ensure initial"
+                                   " variable values\n\tdo not violate"
+                                   " any domains (are you using log/log10?)"
+                                 % con.name)
```

Comment:
You're right, I forgot to include the ValueError in the except statement.

As for the rest of the exception raising, most information is already included in the logger errors despite my current code raising a new exception. Here is an example of the (long) output if I just do nothing but raise the original exception:

`ERROR: evaluating object as numeric value: log10( x8 )
            (object: <class 'pyomo.core.kernel.expr_coopr3._IntrinsicFunctionExpression'>)
        math domain error
ERROR: evaluating expression: math domain error
            (expression: x10 + log10( x8 ))
ERROR: evaluating object as numeric value: x10 + log10( x8 )
            (object: <class 'pyomo.core.kernel.expr_coopr3._SumExpression'>)
        math domain error
ERROR: evaluating expression: math domain error
            (expression: log10( x10 + log10( x8 ) ))
ERROR: evaluating object as numeric value: log10( x10 + log10( x8 ) )
            (object: <class 'pyomo.core.kernel.expr_coopr3._IntrinsicFunctionExpression'>)
        math domain error
ERROR: evaluating expression: math domain error
            (expression: 2 * x6 * log10( x10 + log10( x8 ) ))
ERROR: evaluating object as numeric value: 2 * x6 * log10( x10 + log10( x8 ) )
            (object: <class 'pyomo.core.kernel.expr_coopr3._ProductExpression'>)
        math domain error
ERROR: evaluating expression: math domain error
            (expression: 2 * x6 * log10( x10 + log10( x8 ) ) + x12)
ERROR: evaluating object as numeric value: 2 * x6 * log10( x10 + log10( x8 ) ) + x12
            (object: <class 'pyomo.core.kernel.expr_coopr3._SumExpression'>)
        math domain error
Traceback (most recent call last):
  File "water_demo.py", line 141, in <module>
    m = main()
  File "water_demo.py", line 107, in main
    m.write('wawa.gms')
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/base/block.py", line 1647, in write
    io_options)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/repn/plugins/gams_writer.py", line 187, in __call__
    holdfixed=holdfixed
  File "/home/gseastre/pyomo/src/pyomo/pyomo/repn/plugins/gams_writer.py", line 243, in _write_model
    value(con.body)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/numvalue.py", line 161, in value
    tmp = numeric(exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 192, in __call__
    exception=exception))
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 791, in _apply_operation
    return sum(c*next(values) for c in self._coef) + self._const
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 791, in <genexpr>
    return sum(c*next(values) for c in self._coef) + self._const
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 201, in _evaluate_arglist
    yield value(arg, exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/numvalue.py", line 161, in value
    tmp = numeric(exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 693, in __call__
    exception=exception):
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 201, in _evaluate_arglist
    yield value(arg, exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/numvalue.py", line 161, in value
    tmp = numeric(exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 192, in __call__
    exception=exception))
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 295, in _apply_operation
    return self._operator(*tuple(values))
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 201, in _evaluate_arglist
    yield value(arg, exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/numvalue.py", line 161, in value
    tmp = numeric(exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 192, in __call__
    exception=exception))
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 791, in _apply_operation
    return sum(c*next(values) for c in self._coef) + self._const
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 791, in <genexpr>
    return sum(c*next(values) for c in self._coef) + self._const
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 201, in _evaluate_arglist
    yield value(arg, exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/numvalue.py", line 161, in value
    tmp = numeric(exception=exception)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 192, in __call__
    exception=exception))
  File "/home/gseastre/pyomo/src/pyomo/pyomo/core/kernel/expr_coopr3.py", line 295, in _apply_operation
    return self._operator(*tuple(values))
ValueError: math domain error`

As you can see, the logger error lines already capture the original exception message. Also, the traceback here is quite large, depending on the depth of the violating expression. That being said, I don't personally know which course here is preferred. Is it preferred to keep this long traceback and reraise the original exception (after throwing my own logger error)?